### PR TITLE
[SPARK-33523][SQL][TEST] Add predicate related benchmark to SubExprEliminationBenchmark

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This patch adds predicate related benchmark to `SubExprEliminationBenchmark`.

### Why are the changes needed?

We should have a benchmark for subexpression elimination of predicate.

### Does this PR introduce _any_ user-facing change?

No, dev only.

### How was this patch tested?

Run benchmark locally.

### CODE REVIEW

**Review by maropu — APPROVED** [review]

File: sql/core/benchmarks/SubExprEliminationBenchmark-results.txt [hunk 1]
```diff
@@ -5,11 +5,21 @@ Benchmark for performance of subexpression elimination
...
+subexpressionElimination on, codegen off            2185           2254          64          0.0    21852694.0      11.8X
+
+Preparing data for benchmarking ...
+OpenJDK 64-Bit Server VM 1.8.0_265-b01 on Mac OS X 10.15.6
+Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
+from_json as subExpr in Filter:            Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
+-------------------------------------------------------------------------------------------------------------------------
+subexpressionElimination off, codegen on           42687          42805         111          0.0   426873372.5       1.0X
+subexpressionElimination off, codegen off          43606          45108        1613          0.0   436055236.3       1.0X
+subexpressionElimination on, codegen on            29761          30563         704          0.0   297614324.4       1.4X
```
Comment (dongjoon-hyun) [inline sql/core/benchmarks/SubExprEliminationBenchmark-results.txt:22]: In this case, do we have only 40% improvement?

**Review by dongjoon-hyun — APPROVED** [review]

File: sql/core/benchmarks/SubExprEliminationBenchmark-results.txt (hunk 1 above)
Comment (viirya) [reply sql/core/benchmarks/SubExprEliminationBenchmark-results.txt:22]: Seems so, it is subexpression elimination under codegen.

Comment (viirya) [reply sql/core/benchmarks/SubExprEliminationBenchmark-results.txt:22]: I will submit subexpression elimination under non-codegen after this.

**Comment by viirya** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SubExprEliminationBenchmark.scala [hunk 2]
```diff
@@ -108,11 +110,81 @@ object SubExprEliminationBenchmark extends SqlBasedBenchmark {
...
+      benchmark.addCase("subexpressionElimination off, codegen on", numIters) { _ =>
+        withSQLConf(
+          SQLConf.SUBEXPRESSION_ELIMINATION_ENABLED.key -> "false",
+          SQLConf.WHOLESTAGE_CODEGEN_ENABLED.key -> "true",
+          SQLConf.CODEGEN_FACTORY_MODE.key -> "CODEGEN_ONLY",
+          SQLConf.JSON_EXPRESSION_OPTIMIZATION.key -> "false") {
+          val df = spark.read
+            .text(path.getAbsolutePath)
+            .where(Column(predicate))
+          df.collect()
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SubExprEliminationBenchmark.scala:136]: Can we use `write.format("noop").save()`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SubExprEliminationBenchmark.scala [hunk 3]
```diff
@@ -108,11 +110,81 @@ object SubExprEliminationBenchmark extends SqlBasedBenchmark {
...
+
+      val predicate = (0 until numCols).map { idx =>
+        (from_json('value, schema).getField(s"col$idx") >= Literal(100000)).expr
+      }.asInstanceOf[Seq[Expression]].reduce(Or)
+
+      // We only benchmark subexpression performance under codegen/non-codegen, so disabling
+      // json optimization.
+      benchmark.addCase("subexpressionElimination off, codegen on", numIters) { _ =>
+        withSQLConf(
+          SQLConf.SUBEXPRESSION_ELIMINATION_ENABLED.key -> "false",
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SubExprEliminationBenchmark.scala:129]: Should we maybe better do this with a loop? e.g.)

```scala
Seq(("false", "true", "CODEGEN_ONLY", "false"), ...). foreach {
  case (
       subexpressionElimitationEnabled,
       ...  =>
     benchmark.addCase(s"${SQLConf.key} $subexpressionElimitationEnabled, ...", numIters) { _ =>
        withSQLConf(
          SQLConf.key -> subexpressionElimitationEnabled,
          ...
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SubExprEliminationBenchmark.scala (hunk 2 above)
Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/SubExprEliminationBenchmark.scala]: Okay. I reran the benchmark.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SubExprEliminationBenchmark.scala (hunk 3 above)
Comment (HyukjinKwon) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/SubExprEliminationBenchmark.scala:129]: I believe we can do this separately to refactor this whole file.

**Review by HyukjinKwon — APPROVED** [review]
LGTM Thanks @viirya

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SubExprEliminationBenchmark.scala (hunk 3 above)
Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/SubExprEliminationBenchmark.scala:129]: Okay, I will refactor it.

**Comment by HyukjinKwon** [comment]
GA passed. 
Merged to master.

**Comment by viirya** [comment]
Thanks @HyukjinKwon @dongjoon-hyun @maropu

**Pull request closed without merging**